backend/wmg/data/wmg_cube.py

```diff
@@ -36,3 +48,216 @@ def create_cell_count_cube(tdb_group: str):
+        logger.debug("Cube consolidated, start vacuumming")
+        tiledb.vacuum(uri)
+
+    logger.debug("Cube creation complete")
+    create_cube_sec = time.time() - start_time
+    logger.info(f"Big cube: time to create {create_cube_sec}, uri={uri}")
+
+
+def create_empty_cube(uri: str):
+    """ "
+    Create an empty cube with expected schema (dimensions and attributes) at given uri
+    """
+    if tiledb.array_exists(uri):
+        tiledb.remove(uri)
+
+    tiledb.Array.create(
+        uri,
+        expression_summary_schema,
+    )
+
+
+def load_data_into_cube(tdb_group, uri: str):
+    """
+    Load data from the concatenated corpus into the queryable cube
+    """
+    ctx = create_ctx()
+    start_time = time.time()
+    logger.debug(f"Start loading big cube at : {uri}")
+
+    with tiledb.open(f"{tdb_group}/var", ctx=ctx) as var:
+        gene_ontology_term_ids = var.query(dims=["gene_ontology_term_id"], attrs=[], use_arrow=False).df[:]
+    n_genes = len(gene_ontology_term_ids)
+
+    ##
+    # Reduce X
+    ##
+    big_cube_atts = cube_indexed_dims_no_gene_ontology + cube_non_indexed_dims
+    cell_labels, cube_index = make_cube_index(tdb_group, big_cube_atts)
+    n_groups = len(cube_index)
+
+    cube_sum = np.zeros((n_groups, n_genes), dtype=np.float32)
+    cube_nnz = np.zeros((n_groups, n_genes), dtype=np.uint64)
+    cube_min = np.zeros((n_groups, n_genes), dtype=np.float32)
+    cube_max = np.zeros((n_groups, n_genes), dtype=np.float32)
+
+    # pass 1 - sum, nnz, min, max
+    reduce_X(tdb_group, start_time, cell_labels.cube_idx.values, cube_sum, cube_nnz, cube_min, cube_max)
+
+    return build_in_mem_cube(gene_ontology_term_ids, cube_index, cube_non_indexed_dims, cube_sum, cube_nnz)
+
+
+def build_in_mem_cube(gene_ids, cube_index, other_attrs, cube_sum, cube_nnz):
+    """
+    Build the cube in memory, calculating the
+    """
+    logger.info("Building in-mem cube")
+
+    # Count total values so we can allocate buffers once
+    total_vals = 0
+    for cube_idx in cube_index.cube_idx.values:
+        mask = cube_nnz[cube_idx] != 0
+        total_vals += np.count_nonzero(mask)
+
+    # allocate buffers
+    dims = [
+        np.empty((total_vals,), dtype=object),
+        np.empty((total_vals,), dtype=object),
+        np.empty((total_vals,), dtype=object),
+    ]
+    vals = {
+        "sum": np.empty((total_vals,)),
+        "nnz": np.empty((total_vals,), dtype=np.uint64),
+        "n_cells": np.empty((total_vals,), dtype=np.uint32),
+        **{k: np.empty((total_vals,), dtype=object) for k in other_attrs},
+    }
+
+    # populate buffers
+    idx = 0
+
+    for grp in cube_index.to_records():
+        (
+            tissue_ontology_term_id,
+            organism_ontology_term_id,
+            *attr_values,
+            n,
+            cube_idx,
+        ) = grp.tolist()
+        mask = cube_nnz[cube_idx] != 0
+        n_vals = np.count_nonzero(mask)
+        if n_vals == 0:  # Todo what is the purpose of this?
```

Comment:
re:TODO, to maintain sparsity, presumably, if we want to update comment to that effect

Reply:
thanks